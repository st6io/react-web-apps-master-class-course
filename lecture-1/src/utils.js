import { get } from 'lodash';

export const currency = {
  usd: 'USD',
  bgn: 'BGN',
  eur: 'EUR',
};

export const EUR_TO_BGN_RATE = 1.95583;

const random = (min, max) => Math.random() * (max - min) + min;

const getCurrencies = () => {
  const usdToBgn = random(1.6564, 1.7923, 4);
  const usdToEur = usdToBgn / 1.95583;

  return {
    [currency.usd]: {
      [currency.bgn]: usdToBgn,
      [currency.eur]: usdToEur,
      [currency.usd]: 1,
    },
    [currency.bgn]: {
      [currency.usd]: 1 / usdToBgn,
      [currency.eur]: 1 / EUR_TO_BGN_RATE,
      [currency.bgn]: 1,
    },
    [currency.eur]: {
      [currency.usd]: 1 / usdToEur,
      [currency.bgn]: EUR_TO_BGN_RATE,
      [currency.eur]: 1,
    },
  };
};

const currentCurrencies = getCurrencies();

export const toFixed = (value, precision = 2) => {
  const f = 10 ** precision;

  return Math.round(value * f) / f;
};

export const convertCurrency = (
  value,
  currency,
  targetCurrency,
  precision = 2,
) => {
  const input = parseFloat(value);
  if (Number.isNaN(input)) {
    return '';
  }

  const ratio = get(currentCurrencies, `${currency}.${targetCurrency}`, NaN);
  if (Number.isNaN(ratio)) {
    throw new Error(`cannot convert ${currency} to ${targetCurrency}`);
  }

  const converted = value * ratio;

  return toFixed(converted, precision);
};

const delay = (result, timeout = 250) =>
  new Promise(resolve => setTimeout(() => resolve(result), timeout));

export const fetchTempertature = () => delay('20C', 1000);

export const fetchHumidity = () => delay('54%');
