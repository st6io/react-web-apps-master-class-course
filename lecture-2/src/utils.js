const images = [
  'https://www.maxim.com/.image/t_share/MTYwNjI1MDQ4NTM1NzA0NzQ3/jocko-willink-attack-mondays.png',
  'https://i.ytimg.com/vi/ljqra3BcqWM/maxresdefault.jpg',
  'https://peterattiamd.com/wp-content/uploads/2019/06/jocko-part2-2x1.png',
  'https://i.ytimg.com/vi/Ls5yHTGVzgU/maxresdefault.jpg',
  'https://miro.medium.com/max/3600/0*uvDOX4ea2mGXdAQ9',
];

export const getRandomImage = async () => {
  const response = await fetch(
    'https://qrng.anu.edu.au/API/jsonI.php?length=1&type=uint16',
  );

  const {
    data: [rnd],
  } = await response.json();

  return images[rnd % images.length];
};

export const sum = (a, b) => a + b;
