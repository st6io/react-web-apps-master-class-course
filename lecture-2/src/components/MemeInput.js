import React from 'react';
import Row from './Row';

import styles from './MemeInput.module.scss';
import { getRandomImage } from '../utils';

export default class MemeInput extends React.Component {
  state = {
    showUrl: false,
  };

  toggleUrlInput = () => {
    this.setState(({ showUrl }) => ({ showUrl: !showUrl }));
  };

  onTextChange = ({ target: { name, value } }) => {
    this.props.onInputChange({ [name]: value });
  };

  shuffleImage = async () => {
    const image = await getRandomImage();

    this.props.onImageChange(image);
  };

  render() {
    const { topText, bottomText, url, className } = this.props;

    return (
      <div className={className}>
        <Row>
          <input
            type="text"
            value={topText}
            onChange={this.onTextChange}
            name="topText"
          />
        </Row>
        <Row>
          <input
            type="text"
            value={bottomText}
            onChange={this.onTextChange}
            name="bottomText"
          />
        </Row>

        <Row>
          <div className={styles.leftColumn}>
            <button onClick={this.toggleUrlInput}>Via Url</button>
            <button id="shuffle" onClick={this.shuffleImage}>
              Shuffle
            </button>
          </div>
        </Row>

        {this.state.showUrl && (
          <Row>
            <input
              type="text"
              value={url}
              onChange={this.onTextChange}
              name="url"
            />
          </Row>
        )}
      </div>
    );
  }
}
