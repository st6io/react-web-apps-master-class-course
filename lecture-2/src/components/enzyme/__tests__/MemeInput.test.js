import React from 'react';
import { shallow } from 'enzyme';
import { getRandomImage } from '../../../utils';

import MemeInput from '../../MemeInput';

jest.mock('../../../utils', () => ({
  getRandomImage: jest.fn(),
}));

describe('MemeInput Component', () => {
  it('smoke test', () => {
    const wrapper = shallow(<MemeInput />);

    expect(wrapper).toExist();
  });

  it('should match snapshot', () => {
    const wrapper = shallow(
      <MemeInput
        className="new-class-name"
        topText="Your excuses"
        bottomText="are lies!"
        onTextChange={() => {}}
        shuffleImage={() => {}}
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot when url toggled', () => {
    const wrapper = shallow(
      <MemeInput
        className="new-class-name"
        topText="Your excuses"
        bottomText="are lies!"
        onTextChange={() => {}}
        shuffleImage={() => {}}
      />,
    );

    const toggleUrlButton = wrapper.findWhere(
      el => el.text() === 'Via Url' && el.type() === 'button',
    );
    toggleUrlButton.simulate('click');

    expect(wrapper).toMatchSnapshot();
  });

  it('should render url input when Via Url clicked', () => {
    const wrapper = shallow(
      <MemeInput
        className="new-class-name"
        topText="Your excuses"
        bottomText="are lies!"
        onTextChange={() => {}}
        shuffleImage={() => {}}
      />,
    );

    const toggleUrlButton = wrapper.findWhere(
      el => el.text() === 'Via Url' && el.type() === 'button',
    );

    toggleUrlButton.simulate('click');

    expect(wrapper.find('input[name="url"]')).toExist();
  });

  it('should propagate topText change', () => {
    const onInputChangeMock = jest.fn();
    const wrapper = shallow(
      <MemeInput
        className="new-class-name"
        topText="Your excuses"
        bottomText="are lies!"
        onInputChange={onInputChangeMock}
        shuffleImage={() => {}}
      />,
    );

    expect(onInputChangeMock).not.toHaveBeenCalled();

    const topTextInput = wrapper.find('input[name="topText"]');
    topTextInput.simulate('change', {
      target: { name: 'topText', value: 'test test test' },
    });

    expect(onInputChangeMock).toHaveBeenCalledWith({
      topText: 'test test test',
    });
  });

  it('should propagate image change', async () => {
    const expectedImage = 'https://some.test.img/png';

    const resolveExpectedImage = Promise.resolve(expectedImage);
    getRandomImage.mockImplementation(() => resolveExpectedImage);

    const onImageChangeMock = jest.fn();
    const wrapper = shallow(
      <MemeInput
        className="new-class-name"
        topText="Your excuses"
        bottomText="are lies!"
        onImageChange={onImageChangeMock}
      />,
    );

    expect(onImageChangeMock).not.toHaveBeenCalled();

    const shuffleButton = wrapper.findWhere(
      el => el.text() === 'Shuffle' && el.type() === 'button',
    );

    shuffleButton.simulate('click');

    await resolveExpectedImage;

    expect(onImageChangeMock).toHaveBeenCalledWith(expectedImage);
  });
});
