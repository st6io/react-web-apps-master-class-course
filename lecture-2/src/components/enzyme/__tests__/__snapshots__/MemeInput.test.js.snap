// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemeInput Component should match snapshot 1`] = `
<div
  className="new-class-name"
>
  <Row>
    <input
      name="topText"
      onChange={[Function]}
      type="text"
      value="Your excuses"
    />
  </Row>
  <Row>
    <input
      name="bottomText"
      onChange={[Function]}
      type="text"
      value="are lies!"
    />
  </Row>
  <Row>
    <div
      className="leftColumn"
    >
      <button
        onClick={[Function]}
      >
        Via Url
      </button>
      <button
        id="shuffle"
        onClick={[Function]}
      >
        Shuffle
      </button>
    </div>
  </Row>
</div>
`;

exports[`MemeInput Component should match snapshot when url toggled 1`] = `
<div
  className="new-class-name"
>
  <Row>
    <input
      name="topText"
      onChange={[Function]}
      type="text"
      value="Your excuses"
    />
  </Row>
  <Row>
    <input
      name="bottomText"
      onChange={[Function]}
      type="text"
      value="are lies!"
    />
  </Row>
  <Row>
    <div
      className="leftColumn"
    >
      <button
        onClick={[Function]}
      >
        Via Url
      </button>
      <button
        id="shuffle"
        onClick={[Function]}
      >
        Shuffle
      </button>
    </div>
  </Row>
  <Row>
    <input
      name="url"
      onChange={[Function]}
      type="text"
    />
  </Row>
</div>
`;
